File: Leetcode/Algorithm/cpp/Maximum Length of Pair Chain.cc
const int N = 2222;

class Solution {
public:
    int findLongestChain(vector<vector<int>>& pairs) {
        int n = pairs.size();
        vector<int> dp(n, 0);
        
        sort(pairs.begin(), pairs.end(), [](const vector<int>& va, const vector<int>& vb) {
            return va[1] < vb[1];
        });
        
        int ans = 0;
        for (int i = 0; i < n; i++) {
            dp[i] = 1;
            for (int j = 0; j < i; j++) {
                if (pairs[i][0] > pairs[j][1]) {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
            ans = max(ans, dp[i]);
        }
        
        return ans;
    }
};

File: Leetcode/Algorithm/python/Maximum Length of Pair Chain.py
INF = 10 ** 10

class Solution(object):
    def findLongestChain(self, pairs):
        pre, cnt = -INF, 0
        for pair in sorted(pairs, key=lambda x: x[1]):
            (l, r) = pair
            if l > pre:
                pre = r
                cnt += 1
        return cnt
        
